akd/src/directory.rs

```diff
@@ -233,68 +244,155 @@ impl<S: Storage + Sync + Send, V: VRFKeyStorage> Directory<S, V> {
+                .get_label_proof::<H>(&uname, false, current_version)
+                .await?
+                .to_bytes()
+                .to_vec(),
+            existence_proof: current_azks
+                .get_membership_proof(&self.storage, lookup_info.existent_label, current_epoch)
+                .await?,
+            marker_vrf_proof: self
+                .vrf
+                .get_label_proof::<H>(&uname, false, lookup_info.marker_version)
+                .await?
+                .to_bytes()
+                .to_vec(),
+            marker_proof: current_azks
+                .get_membership_proof(&self.storage, lookup_info.marker_label, current_epoch)
+                .await?,
+            freshness_vrf_proof: self
+                .vrf
+                .get_label_proof::<H>(&uname, true, current_version)
+                .await?
+                .to_bytes()
+                .to_vec(),
+            freshness_proof: current_azks
+                .get_non_membership_proof(
+                    &self.storage,
+                    lookup_info.non_existent_label,
+                    current_epoch,
+                )
+                .await?,
+        };
+
+        // self.storage.log_metrics(log::Level::Info).await;
+
+        Ok(lookup_proof)
+    }
+
+    // TODO(eoz): Call proof generations async
+    /// Allows efficient batch lookups by preloading necessary nodes for the lookups.
+    pub async fn batch_lookup<H: Hasher>(
+        &self,
+        unames: &[AkdLabel],
+    ) -> Result<Vec<LookupProof<H>>, AkdError> {
+        let current_azks = self.retrieve_current_azks().await?;
+        let current_epoch = current_azks.get_latest_epoch();
+
+        // Take a union of the labels we will need proofs of for each lookup.
+        let mut lookup_labels = Vec::new();
+        let mut lookup_infos = Vec::new();
+        for uname in unames {
+            // Save lookup info for later use.
+            let lookup_info = self
+                .get_lookup_info::<H>(uname.clone(), current_epoch)
+                .await?;
+            lookup_infos.push(lookup_info.clone());
+
+            // A lookup proofs consists of the proofs for the following labels.
+            lookup_labels.push(lookup_info.existent_label);
+            lookup_labels.push(lookup_info.marker_label);
+            lookup_labels.push(lookup_info.non_existent_label);
+        }
+
+        // Create a union of set of prefixes we will need for lookups.
+        let lookup_prefixes_set = crate::utils::build_lookup_prefixes_set(&lookup_labels);
+
+        // Load nodes.
+        current_azks
+            .bfs_preload_nodes::<_, H>(&self.storage, lookup_prefixes_set)
+            .await?;
+
+        // Ensure we have got all lookup infos needed.
+        assert_eq!(unames.len(), lookup_infos.len());
+
+        let mut lookup_proofs = Vec::new();
+        for i in 0..unames.len() {
+            lookup_proofs.push(
+                self.lookup_with_info::<H>(
```

Comment:
Why do we need the info? I think if we've done the proper BFS loading, we should only need to batch-get the value states in order to load them into the cache, then we can do the regular proof generation operations since we'll only be accessing from local memory

Reply:
This is to eliminate duplicate `get_user_state` queries. One in identifying which labels are needed and one in (previously) lookup for the same purpose. This way pre-loaded info is passed to the lookup function, so no re-loads.

Looking at the code for MySQL's `get_user_state`, are we utilizing the cache?

Reply:
Ah I see, yes we've excluded `get_user_state` from the caching logic because of the filter parameters. How would you know a cached value matches "<= epoch" or some weird filter without going to the backing store? It sounds like we might need a new query at the storage layer to retrieve the max versions for a collection of users which is <= a given epoch. (If I'm remembering right). We could move forward with this however and simply open an issue for it, but we'll want to do it properly for batch proof generation. Here's we will hit order N queries just to gather the user states, while not awful, still not ideal when we can do a better query.

Reply:
Ah I see! Tracked in #169. I can take a stab at #166, #167 and #169 in a new PR. Does that sound good?

Reply:
Sounds great, let's merge this in!